### PR TITLE
Add patient features in cash penalty env

### PR DESCRIPTION
**Pull request by louisowen6**
Add `patient` parameter to control whether we have to end the cycle when we're running out of cash or just don't buy anything until we got additional cash

### CODE REVIEW

**Review by spencerR1992 — CHANGES_REQUESTED** [review]
Please add testing for this feature to verify correct operation!  Thanks

File: finrl/env/env_stocktrading_cashpenalty.py [hunk 1]
```diff
@@ -288,10 +288,17 @@ def step(self, actions):
...
+                    self.log_step(reason="CASH SHORTAGE")
+                    actions = np.where(actions>0,0,actions)
+                    spend = 0
+                    costs = 0
+                else:
+                    # ... end the cycle and penalize
+                    return self.return_terminal(
+                        reason="CASH SHORTAGE",reward=self.get_reward()
+                    )
             self.transaction_memory.append(actions)  # capture what the model's could do
```
Comment (spencerR1992) [inline finrl/env/env_stocktrading_cashpenalty.py:302]: This should be before any modifications to action, so please put this higher in the step function. Actions == raw decisions from model. Transactions=  actual actions applied in this step.

Comment (louisowen6) [reply finrl/env/env_stocktrading_cashpenalty.py:302]: I think that's transaction_memory not action_memory, so it should be okay to put it down there to also capture the changes of actions if the user turns on the patient parameter. What do you think?

**Comment by louisowen6** [comment]
Added test module for `patient` parameter

File: finrl/env/env_stocktrading_cashpenalty.py (hunk 1 above)
Comment (spencerR1992) [reply finrl/env/env_stocktrading_cashpenalty.py:302]: oh, yes, sorry you're right!

File: tests/test_env/test_env_cashpenalty.py [hunk 2]
```diff
@@ -67,6 +67,23 @@ def test_shares_increment(self):
         self.assertEqual(holdings[0], 10.0)
         self.assertEqual(holdings[1], 0.0)
 
+    def test_patient(self):
```
Comment (spencerR1992) [inline tests/test_env/test_env_cashpenalty.py:70]: can you please add test for the opposite case ? When `patient=False` then is_done==True`?

**Review by spencerR1992 — APPROVED** [review]

**Comment by sunnyswag** [comment]
hello, I set `patient = True`, but I still can not get into the second `update` state and `Last Date` state.
here are my outputs.
image
<br>
As I understand it, if I set `patient = True`, I still will log the `CASH SHOTAGE` state.
But, I will come into the `update` state 、second `update` state and `Last Date` state, 
The output is not what I expected.
I can't find the solution, can you give me some advice, please?

**Comment by sunnyswag** [comment]
Oh, I find the reason.
I edit the `patient = True` in FinRL directory, but I import env from python packege.
The solution is set `patient = True` in jupyter notebook.

**Pull request merged**